jax/scipy/sparse/linalg.py

```diff
@@ -167,10 +183,254 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
 
   cg_solve = partial(
       _cg_solve, x0=x0, tol=tol, atol=atol, maxiter=maxiter, M=M)
+
   # real-valued positive-definite linear operators are symmetric
-  real_valued = lambda x: not issubclass(x.dtype.type, np.complexfloating)
+  def real_valued(x):
+    return not issubclass(x.dtype.dtype, np.complexfloating)
   symmetric = all(map(real_valued, tree_leaves(b)))
   x = lax.custom_linear_solve(
       A, b, solve=cg_solve, transpose_solve=cg_solve, symmetric=symmetric)
   info = None  # TODO(shoyer): return the real iteration count here
   return x, info
+
+
+def _safe_normalize(x, return_norm=False, thresh=None):
+  norm = jnp.sqrt(_vdot_tree(x, x, assume_real=False))
+  if thresh is None:
+    thresh = jnp.finfo(x.dtype).eps
+  thresh = thresh.astype(norm.dtype)
+
+  normalized_x, norm = lax.cond(
+    norm > thresh,
+    lambda y: (_div(y, norm), norm),
+    lambda y: (_mul(0.*y, thresh), thresh),  # To get the dtype right
+    x,
+  )
+  if return_norm:
+    return normalized_x, norm
+  else:
+    return normalized_x
+
+
+def kth_arnoldi_iteration(k, A, M, V, H, tol):
+  """
+  Performs a single (the k'th) step of the Arnoldi process. Thus,
+  adds a new orthonormalized Krylov vector A(M(V[:, k])) to V[:, k+1],
+  and that vectors overlaps with the existing Krylov vectors to
+  H[k, :]. The tolerance 'tol' sets the threshold at which an invariant
+  subspace is declared to have been found, in which case the new
+  vector is taken to be the zero vector.
+  """
+
+  v = tree_map(lambda x: x[..., k], V)  # Gets V[:, k]
+  v = A(M(v))
+
+  def gram_schmidt_step(r, v_i):
+    h_i = _vdot_tree(r, v_i)
+    r_i = _sub(r, _mul(h_i, v_i))
+    return r_i, h_i
+  v, h = lax.scan(gram_schmidt_step, v, xs=V.T)
```

Comment:
It looks like this scan is scanning over the columns of `V`?

One key implementation detail from https://github.com/google/jax/pull/4025 / https://github.com/google/jax/issues/1531#issuecomment-657164860 is the use of a matrix/vector product rather than a loop. As I noted in that comment, "This is perhaps 10-20% faster on CPU and ~10x faster on GPUs". So I suspect this is missing that aspect.

My version also used two iterations of classical gram Schmidt, which apparently can be important for ill-conditioned matrices per https://slepc.upv.es/documentation/reports/str1.pdf

Reply:
I had not realized this was so carefully considered, and will restore the old code.